add spherical function

davydden opened the pull request:
this is a WIP of simplifying usage of functions given in spherical coordinates.

An alternative would be to instead implement
```
Tensor<1,dim> GeometricUtilities::Coordinates::from_spherical (const std_cxx11::array< double, dim > & 	scoord,
const std_cxx11::array<double, dim> & sgradient)
```
which transforms gradient from spherical to cartesian coordinates.

However there are a few things that have to be re-used, especially when implementing Hessian. So maybe it's better to keep a separate function which later may not even rely on `GeometricUtilities::Coordinates` to make things more efficient.

**TODO**:
- [x] hessian
- [x] unit tests
- [x] don't use `SymmetricTensor`
- [x] change log

bangerth commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,153 @@
+  namespace
+  {
+    template <int dim>
+    Tensor<1,dim> to_tensor(const Point<dim> &p)
```
I don't think you need this -- `Point<dim>` is already derived from `Tensor<1,dim>`. If you want to make the conversion explicit, just use a `static_cast`.

bangerth commented on include/deal.II/base/geometric_utilities.h:
```diff
@@ -45,16 +45,30 @@ namespace GeometricUtilities
+     *  r &= \sqrt{x^2+y^2+z^2} \\
+     *  \theta &= {\rm atan}(y/x) \\
+     *  \phi &= {\rm acos} (z/r)
+     * @f}
```
Can you submit the changes in this file separately? They are independent and uncontroversial.

davydden replied:
see https://github.com/dealii/dealii/pull/3858

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,97 @@
+
+  /**
+   * The scalar-valued function given in spherical coordinates. The convention
+   * for angles is the same as in GeometricUtilities::Coordinates.
```
I don't think I understand what this class does. Can you elaborate? Maybe with formulas?

davydden replied:
@bangerth i updated the docs with formula and more explanations. Hopefully it's clearer now.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,97 @@
+   *
+   * This is an abstract class
+   * which wrapps transformation of values and gradients from spherical
+   * coordinates to cartesian.
```
-> Cartesian

bangerth commented:
Oh, so you mean that a derived class implements the derivatives in spherical coordinates, and this class then converts them to the usual convention?

davydden commented:
> Oh, so you mean that a derived class implements the derivatives in spherical coordinates, and this class then converts them to the usual convention?

yes, of course values are trivial, gradients are ok, but hessians are nasty. Thus this class.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Gradient in spherical coordinates
+     */
+    virtual Tensor<1,dim> sgradient(const std_cxx11::array<double, dim> &sp) const = 0;
```
I don't think you can represent the gradient in spherical coordinates as a tensor, for the same reason as you can't represent the point itself as a tensor. Or do you mean to say that the point is in spherical coordinates, but the return value in Cartesian?

davydden replied:
this is not a tensor from math point of view, but it's a collection of three numbers. It just simplifies the internal usage. Return value is `[df/dr, df/d\theta, df/d\phi]`

bangerth replied:
I've really taken to think that just because something is *convenient* to represent one way doesn't necessarily make it *useful* in the long run. In particular, I have over the years really come to think that *readability* of code is a really important thing, and that means that we should use `Tensor` only for things that really *are* tensors.

In your case, I would suggest using `array` for the return type, or to invent a new class.

davydden replied:
@bangerth I would use `array` if i would not need to do hessians. IMHO it simply does not make sense to invert yet another class to store a symmetric matrix which is, however, not a tensor.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Hessian in spherical coordinates
+     */
+    virtual SymmetricTensor<2,dim> shessian (const std_cxx11::array<double, dim> &sp) const = 0;
```
same here, obviously

davydden replied:
same as above, obviously ;-)

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,100 @@
+     *
+     * Note: it is stored as a tensor so that operator- naturally returns a point.
+     */
+    const Tensor<1,dim> center;
```
That seems wrong -- the center is a point; if the difference doesn't have the correct type, then use a cast where necessary.

Alternatively, you may want to see the center as an "offset", and in that case the type is correct but one may want to change the name?

davydden replied:
yes, you can consider is as an offset from `0` to some point `p`. The reason for using `Tensor` is that each evaluation starts with 
```
const Point<dim> point = p_ - center
```
where `p_` is the input Point.

bangerth replied:
So maybe call it `coordinate_system_offset` and leave it as a tensor?

jppelteret commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,179 @@
+    if (sg[1] != 0.)
+      {
+        Assert (sp[0] > 0.,
+                ExcDivideByZero());
```
I get that `|r|>0`, but is `ExcDivideByZero()` the best way to describe an error when `sp[0]<0`?

davydden replied:
`|r|>=0`, so yes, `ExcDivideByZero` should be fine to describe the case `r==0` :smile:

jppelteret replied:
But `sp` is a `std_cxx11::array<double, dim>`, so `sp[0]` **could** be negative (although one would expect that `GeometricUtilities::Coordinates::to_spherical()` would check that this is not the case).

davydden replied:
that's exactly the point, `GeometricUtilities::Coordinates::to_spherical()` will never return negative value in first element of the array.

jppelteret commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,109 @@
+   * spherical coordiantes (see svalue(), sgradient() and
+   * shessian() ).
+   *
+   * @note This function is currently only implemented for dim==3 .
```
Surely spherical coordinates are strictly 3d only?

davydden replied:
it depends, `GeometricUtilities` for convenience also do it in 2d where one simply drops elevation angle.

jppelteret commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,179 @@
+    const double sin_phi   = std::sin(sp[2]);
+
+    Tensor<1,dim> unit_r, unit_theta, unit_phi;
+    set_unit_vectors(unit_r, unit_theta, unit_phi,
```
I'd be tempted to send `sp` into this function and just recompute all of `cos_theta` etc. again

davydden replied:
i would prefer not to do extra 4 sin/cosine operations.

jppelteret commented on tests/base/functions_spherical_01.cc:
```diff
@@ -0,0 +1,163 @@
+};
+
+
+// same as above but using Funcitons::Spherical
```
typo: Functions

jppelteret commented on tests/base/functions_spherical_01.cc:
```diff
@@ -0,0 +1,163 @@
+                        ExcInternalError());
+
+          // check hessian:
+          // TODO
```
👍

davydden commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,109 @@
+    virtual double value (const Point<dim> &point,
+                          const unsigned int component = 0) const;
+
+    virtual Tensor<1,dim> gradient (const Point<dim>   &p,
```
i would certainly not add formulas, as for `hessian`'s it's a nightmare. Otherwise the doxygen will be take from the parent class, so i don't think there is a need to add anything here.

davydden commented:
@bangerth @jppelteret i addressed most of your comments, this PR is ready for re-review. I will add another unit test later...

jppelteret commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+    const std_cxx11::array<double, dim> sp = GeometricUtilities::Coordinates::to_spherical(p);
+    const Tensor<1,dim> sg = sgradient(sp);
+
+    // FIXME: somewhat backwards, but we need cos/sin's for unit vectors
```
I suppose that if you're not going to address this now, then its probably best to remove this comment. Same goes for the comment in the Hessian calculation.

jppelteret commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+    const double c_utheta_ur   = r*sh[0][1]*sin_phi-sg[1]*sin_phi;
+    const double c_utheta_uphi = sh[1][2]*sin_phi-cos_phi*sg[1];
+
+    // const double c_ur_utheta   = (r*sh[0][1]-sg[1])                 * sin_phi;
```
Can you remove the commented out lines? Both here and below.

jppelteret commented on tests/base/functions_spherical_01.cc:
```diff
@@ -0,0 +1,167 @@
+          AssertThrow ( dh.norm() <= h1.norm() * 1e-10,
+                        ExcInternalError());
+        }
+  deallog << "Ok"<< std::endl;
```
To be particularly pedantic, I think that we typically write `"OK"` :-)

jppelteret commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+  Spherical<dim>::value (const Point<dim>   &p_,
+                         const unsigned int) const
+  {
+    AssertThrow(dim==3,
```
Perhaps this assert should go (or be duplicated) in the constructor, so the error is thrown as early as possible.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,122 @@
+  /**
+   * An abstract base class for a scalar-valued function $f=f(r,\theta,\phi)$
+   * defined in spherical coordinates. This class wraps transformation of values,
+   * gradients and hessians from spherical coordinates to Cartesian coordinate system.
```
-> to THE Cartesian coordinate system used by the Function base class.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,122 @@
+     * Constructor which should be provided with @p center defining the origin
+     * of the coordinate system.
+     */
+    Spherical(const Point<dim> &center);
```
Maybe give the center a default value equal to `Point<dim>()`?

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Return the value of the function at the given point.
+     */
+    virtual double value (const Point<dim> &point,
```
maybe say "This function converts the given point to spherical coordinates, calls svalue() with it, and converts the result into Cartesian coordinates."

Same for the other two functions.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Return the value at point @p sp.
+     */
+    virtual double svalue(const std_cxx11::array<double, dim> &sp) const = 0;
```
say something like "Here, @p sp is provided in spherical coordinates."

By the way, why not add the `component` argument the `value()` function has?

davydden replied:
because this function is scalar-valued only.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,122 @@
+                          Tensor<1,dim> &unit_theta,
+                          Tensor<1,dim> &unit_phi,
+                          const double &cos_theta,
+                          const double &sin_theta,
```
put the input arguments first.

That said, this class has no state (= member variables), so I suspect that the function could be `static`. If so, it may also be moved into an anonymous namespace in the `.cc` file, right?

bangerth commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+  // https://github.com/apache/commons-math/blob/master/src/main/java/org/apache/commons/math4/geometry/euclidean/threed/SphericalCoordinates.java
+  // http://mathworld.wolfram.com/SphericalCoordinates.html
+
+  /*derivation of Hessian in Maxima as function fo tensor products of unit vectors:
```
fo -> of

bangerth commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+    template <int dim>
+    void symm_outer_product(SymmetricTensor<2,dim> &out,
+                            const Tensor<1,dim> &in1,
+                            const Tensor<1,dim> &in2)
```
We now have pretty good move-support in compilers, so maybe just *return* the `out` argument, rather than passing it through a reference argument. same for the other functions below.

bangerth replied:
By the way, that's a funny name because the function is not the same as `symmetrize(outer_product(in1, in2))`. The latter will divide the result by two.

bangerth commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+  template <int dim>
+  void
+  Spherical<dim>::set_unit_vectors(Tensor<1,dim> &unit_r,
+                                   Tensor<1,dim> &unit_theta,
```
Yes, looking at the implementation, I think the function may just go into an anonymous namespace.

bangerth commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,303 @@
+  std::size_t
+  Spherical<dim>::memory_consumption () const
+  {
+    return coordinate_system_offset.memory_consumption();
```
That's not quite correct because it neglects the memory consumption of the base class. I think the correct return value is `sizeof(Spherical<dim>)`.

bangerth commented:
I'm still a bit uneasy about using `Tensor` and `SymmetricTensor` as the return types of `sgradient` and `shessian`. That's because -- while convenient -- you cannot meaningfully call any of the member functions of these objects. That suggests that the returned objects can conveniently be stored in the member variables of these tensor classes, but that *they are not tensors*.

If you do want to go this route, I would appreciate if you could add a paragraph to the documentation of these functions that states the fact that the returned values are only represented *for convenience* by these tensor objects, but are not in fact tensors.

Alternatively, how about just doing
```
  typedef std_cxx11::array<...> GradientInSphericalCoordinates;
  typedef std_cxx11::array<std_cxx11::array<...> > HessianInSphericalCoordinates;
```
and then use these types as return types?

bangerth commented:
Oh, two more comments:

* You want the new virtual functions in the `public:` section of the class. Otherwise, they cannot be called from the outside.
* The `Function` class implements `value()`, `gradient()`, and `hessian()` with a body that simply throws an exception. That's useful because it means that a derived class doesn't have to overload *all three* of these functions, but only those that are really used in the context where the class is used. Maybe do the same here?

davydden commented:
@bangerth 

> You want the new virtual functions in the public: section of the class. Otherwise, they cannot be called from the outside.

they are not to be called from the outside. They are only to be called from the inside. The current layout works (see unit tests).

davydden commented:
@bangerth @jppelteret i updated the PR and addressed all your comments. From my perspective, should be ready (of course i will squash prior to merging). Added tests pass for me on macOS with clang and doxygen documentation also renders reasonably.

bangerth commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,123 @@
+     * Constructor which should be provided with @p center defining the origin
+     * of the coordinate system.
+     */
+    Spherical(const Point<dim> center = Point<dim>());
```
Not a big deal one way or the other, but we typically pass points by reference.

davydden replied:
so `(const Point<dim> &center = Point<dim>())` would also work? I somewhat thought that the default value can only be when passing by value.

bangerth replied:
Yes, const reference also works (as well as by value). Only non-const reference doesn't work.

bangerth commented on source/base/function_spherical.cc:
```diff
@@ -0,0 +1,341 @@
+  template <int dim>
+  Spherical<dim>::~Spherical()
+  {
+  }
```
You don't actually need the destructor. It is automatically virtual because the one in the base class already is.

jppelteret commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,118 @@
+                          const unsigned int component = 0) const;
+
+    /**
+     * Return the gradient in Cartesian coordinates at point @p p.
```
"gradient with respect to the Cartesian coordinates"?

jppelteret commented on include/deal.II/base/function_spherical.h:
```diff
@@ -0,0 +1,118 @@
+                                    const unsigned int  component = 0) const;
+
+    /**
+     * Return the Hessian in Cartesian coordinates at point @p p.
```
"Hessian with respect to the Cartesian coordinates"?
